runtime/src/serde_utils.rs

```diff
@@ -1,64 +1,670 @@
+                    .as_ref()
+                    .join(&append_vec_relative_path);
+                let target = local_dir.join(append_vec_abs_path.file_name().unwrap());
+                if std::fs::rename(append_vec_abs_path.clone(), target).is_err() {
+                    let mut copy_options = CopyOptions::new();
+                    copy_options.overwrite = true;
+                    let e = fs_extra::move_items(
+                        &vec![&append_vec_abs_path],
+                        &local_dir,
+                        &copy_options,
+                    )
+                    .map_err(|e| {
+                        format!(
+                            "unable to move {:?} to {:?}: {}",
+                            append_vec_abs_path, local_dir, e
+                        )
+                    })
+                    .map_err(accountsdb_to_io_error);
+                    if e.is_err() {
+                        info!("{:?}", e);
+                        continue;
+                    }
+                };
+
+                // Notify the AppendVec of the new file location
+                let local_path = local_dir.join(append_vec_relative_path);
+                let mut u_storage_entry = Arc::try_unwrap(storage_entry).unwrap();
+                u_storage_entry
+                    .set_file(local_path)
+                    .map_err(accountsdb_to_io_error)?;
+                new_slot_storage.insert(id, Arc::new(u_storage_entry));
+            }
+            Ok((slot, new_slot_storage))
+        })
+        .collect();
+
+    let new_storage_map = new_storage_map?;
+    let mut storage = AccountStorage(new_storage_map);
+
+    // discard any slots with no storage entries
+    // this can happen if a non-root slot was serialized
+    // but non-root stores should not be included in the snapshot
+    storage.0.retain(|_slot, stores| !stores.is_empty());
+
+    let version: u64 = deserialize_from(&mut stream)
+        .map_err(|e| format!("write version deserialize error: {}", e.to_string()))
+        .map_err(accountsdb_to_io_error)?;
+
+    let (slot, bank_hash): (Slot, BankHashInfo) = deserialize_from(&mut stream)
+        .map_err(|e| format!("bank hashes deserialize error: {}", e.to_string()))
+        .map_err(accountsdb_to_io_error)?;
+    accounts_db
+        .bank_hashes
+        .write()
+        .unwrap()
+        .insert(slot, bank_hash);
+
+    // Process deserialized data, set necessary fields in self
+    let max_id: usize = *storage
+        .0
+        .values()
+        .flat_map(HashMap::keys)
+        .max()
+        .expect("At least one storage entry must exist from deserializing stream");
+
+    {
+        let mut stores = accounts_db.storage.write().unwrap();
+        stores.0.extend(storage.0);
+    }
+
+    accounts_db.next_id.store(max_id + 1, Ordering::Relaxed);
+    accounts_db
+        .write_version
+        .fetch_add(version, Ordering::Relaxed);
+    accounts_db.generate_index();
```

Comment:
This isn't related to serialization. We rather want to concentrate on serialization in `serde_utils.rs` here. Could you move this [there](https://github.com/solana-labs/solana/pull/9980/files#r426580593)?

By the way, `generate_index` is another very important (and monster) function, so I want to make this to stand out instead of being berried deep down serialization code.

Yeah, original code was rather spaghetti... Thanks for refactoring, in advance!

Reply:
> That said the calls to generate_index and freeze_accounts can be hoisted out of those two serde_utils functions as they're at the end - I assume this is what you're asking?

Yeah, thanks for catching up a context! :)

Reply:
> This isn't related to serialization.

Ah, also, I didn't refer `write_version` with `this`. I only referred to `generate_index()`.

> I should point out version is read from the stream in AccountsDB::accounts_from_stream aka context_accountsdb_from_stream() so needs to be applied ...

And you're understanding is perfect! :)